### ISSUE TITLE
implement timestep handling for remake

### ISSUE THREAD

**Issue by aappling-usgs** [USGS-VIZLAB/vizlab#288]
#281 starts in on a remake version of `createMakefiles` but doesn't yet do timestamp checking for the fetch section. Let's add this and test the heck out of it.

Because remake doesn't have symbolic targets that get run every time, we need some different way to get remake to re-run. I think the solution is going to be a vizlab wrapper to `remake::make()` that first checks for and deletes any fetch targets that aren't supposed to live forever, THEN runs `remake::make()`. A wrapper function also allows us to add some additional fun functionality; here's my working outline of that function:
```r
vizmake <- function(target_name, ...) {
  
  stop('not yet implemented')
  
  # check the format of the viz.yaml (limited checks so far)
  validateVizYaml()
  
  # decide whether createRemakefile needs to be rerun
  
  # remake::delete any timestamps or fetch items whose time is up

  # run remake::make(target_name, ..., remake_file='remake.yaml'). show a timestamp for #283
  
}
```

**Comment by aappling-usgs** [comment]
![rplot](https://user-images.githubusercontent.com/12039957/31844355-fc6c88f4-b5ac-11e7-8f22-7678195e9f70.png)

```r
DiagrammeR::grViz("
digraph G {

graph [layout = dot,
       rankdir = TB]

node [shape = box,
      style = filled,
      color = grey]

node [fillcolor = pink]
file_exists

node [fillcolor = skyblue]
currency1; currency2

node [fillcolor = yellow]
check_timetolive

node [fillcolor = orange]
fetchTimestamp

node [fillcolor = green]
make_file

node [fillcolor = red]
dont_make_file

node [shape = oval,
      fillcolor = lightgrey]

edge [color = grey]

file_exists -> no -> make_file
file_exists -> yes -> currency1
currency1 -> alwaysCurrent -> dont_make_file
currency1 -> other -> check_timetolive
check_timetolive -> before_timetolive -> dont_make_file
check_timetolive -> after_timetolive -> currency2
currency2 -> sometimesCurrent -> fetchTimestamp
currency2 -> neverCurrent -> make_file
fetchTimestamp -> same_timestamp -> dont_make_file
fetchTimestamp -> new_timestamp -> make_file

{rank = same; no; yes;}
{rank = same; other; alwaysCurrent;}
{rank = same; after_timetolive; before_timetolive;}
{rank = same; sometimesCurrent; neverCurrent;}
{rank = same; new_timestamp; same_timestamp;}
{rank = same; make_file; dont_make_file;}

}
")
```

**Comment by aappling-usgs** [comment]
Here's how we can handle all these cases within remake:

A. file_exists isn't a function, just the state of the repository. remake figures this out.
B. currency1 is an if statement in `vizmake` that decides whether to call  `checkTimeToLive` or not based on whether the item is alwaysCurrent or not (this info is in the viz.yaml).
C. `checkTimeToLive` should compare the file timestamps on the remake 'objects' (.rds files in the store) to the current time (Sys.time()). If the time difference between those two timestamps is greater than TTL, then `remake::delete()` the timestamp file. 
D. currency2 is actually a call to fetchTimestamp every time; i'll revise the graph below to reflect this. The two possibilities beneath that decision point are that fetchTimestamp on a neverCurrent item _always_ creates a new_timestamp, whereas other fetchTimestamp methods should always return a new object (i'd better check that the file timestamps on remake's .rds files get changed every time they get deleted and rebuilt - we need them to) but may return either a new value (if the remote timestamp has changed) or an identical value (if the remote timestamp hasn't changed).
E. if the timestamp object gets changed, remake should figure it out and refetch the file.

The above preprocessing steps should happen on all viz id targets in the remake file each time `vizmake` is called, even if only one or two targets have been called out in `target_names`. thus, calling vizmake on a single process item will still involve re-fetching a dependency of that process item if the fetch item has gone out of date. if someone really wants to bypass this, they should call `process('vizid')` instead of `vizmake('vizid')`.

The above graph has six routes to the bottom. From leftmost (1) to rightmost (6), here's how the above rules play out:

1. Every remake recipe already does this, because remake rules say that if the target file or object doesn't exist, it gets made. Most targets (except those in route 6, the alwaysCurrents) should depend on a fetched timestamp, which remake will create once and then leave untouched as long as we leave it untouched:
    ```r
    cache/cars.csv:
      depends: cars_data_timestamp
      command: fetch('cars-data')
    cars_data_timestamp:
      command: fetchTimestamp('cars-data')
    ``` 
2. When the time to live has been exceeded, never-current fetch items should always have their timestamp objects deleted, and `neverCurrent` should create a new timestamp equal to the current `Sys.time()` so that it's always different from what it was before, so the fetching always happens.
3. When the time to live has been exceeded, sometimes-current fetch items will check their timstamps. If the timestamp has changed, fetching should happen.
4. When the time to live has been exceeded, sometimes-current fetch items will check their timstamps. If the timestamp has NOT changed, fetching should NOT happen.
5. When the time to live has not been exceeded, it doesn't matter whether fetch items are neverCurrent or sometimesCurrent; they are not refetched.
6. Fetch items for which `fetchTimestamp == alwaysCurrent` don't even need to specify a timestamp file or go through the time-to-live check, because they never get refetched:
    ```r
    data/iris.csv:
      command: fetch('iris-data')
    ``` 

With the above strategy in mind, we can revise the graph:
```r
DiagrammeR::grViz("
digraph G {

graph [layout = dot,
       rankdir = TB]

node [shape = box,
      style = filled,
      color = grey]

node [fillcolor = pink]
file_exists

node [fillcolor = skyblue]
currency

node [fillcolor = yellow]
check_timetolive

node [fillcolor = orange]
fetchTimestamp

node [fillcolor = green]
make_file

node [fillcolor = red]
dont_make_file

node [fillcolor = white,
      color = orange]
neverCurrent; sometimesCurrent

node [shape = oval,
      fillcolor = lightgrey,
      color = grey]

edge [color = grey]

file_exists -> no -> make_file
file_exists -> yes -> currency
currency -> alwaysCurrent -> dont_make_file
currency -> other -> check_timetolive
check_timetolive -> before_timetolive -> dont_make_file
check_timetolive -> after_timetolive -> fetchTimestamp
fetchTimestamp -> neverCurrent -> new_timestamp
fetchTimestamp -> sometimesCurrent
sometimesCurrent -> same_timestamp
sometimesCurrent -> new_timestamp
new_timestamp -> make_file
same_timestamp -> dont_make_file

{rank = same; no; yes;}
{rank = same; other; alwaysCurrent;}
{rank = same; after_timetolive; before_timetolive;}
{rank = same; sometimesCurrent; neverCurrent;}
{rank = same; new_timestamp; same_timestamp;}
{rank = same; make_file; dont_make_file;}

}
")
```
![image](https://user-images.githubusercontent.com/12039957/31845020-a6fdfb1e-b5b1-11e7-9f92-6190bbaf8c50.png)

**Comment by aappling-usgs** [comment]
:point_up: `file_exists` should be `file_exists & current_with_deps`

**Comment by aappling-usgs** [comment]
closed by #316
